### PR TITLE
Guidelines: set matchLabels as being mandatory

### PR DESCRIPTION
**Pull request by desaintmartin**
Also move "multiple components" hint and separate the 3 different topics.

See https://github.com/helm/charts/issues/3011 and https://github.com/helm/charts/issues/7680

### CODE REVIEW

File: REVIEW_GUIDELINES.md [hunk 1]
```diff
@@ -44,8 +45,25 @@ labels:
   heritage: {{ .Release.Service }}
 ```
 
+If a chart has multiple components, a `component` label should be added (e. g. `component: server`). The resource name should get the component as suffix (e. g. `name: {{ template "myapp.fullname" . }}-server`).
+
 Note that templates have to be namespaced. With Helm 2.7+, `helm create` does this out-of-the-box. The `app` label should use the `name` template, not `fullname` as is still the case with older charts.
 
+### Deployments and StatefulSets selectors
+spec.selector.matchLabels for `StatefulSets` and `Deployments` must be specified and must have both `app` and `release` labels.
+It should not contain other labels except for edge cases, and especially MUST NOT contain `chart` label or any label containing a version of the chart.
```
Comment (scottrigby) [inline REVIEW_GUIDELINES.md]: We may want to quickly say why not. Something like what you wrote in #7680. I edited that below, as a suggestion:

> This is because the chart label string contains the version, so whenever the the `Chart.yaml` version changes, Helm's attempt to change this immutable field would cause the upgrade to fail.

Comment (desaintmartin) [reply REVIEW_GUIDELINES.md]: Thanks! I've added this.

File: REVIEW_GUIDELINES.md [hunk 2]
```diff
@@ -44,8 +45,29 @@ labels:
   heritage: {{ .Release.Service }}
 ```
 
+If a chart has multiple components, a `component` label should be added (e. g. `component: server`). The resource name should get the component as suffix (e. g. `name: {{ template "myapp.fullname" . }}-server`).
+
 Note that templates have to be namespaced. With Helm 2.7+, `helm create` does this out-of-the-box. The `app` label should use the `name` template, not `fullname` as is still the case with older charts.
 
+In case of a controller like `Deployment` or `StatefulSet`, those labels should also be set in `spec.template.metadata.labels`.
```
Comment (desaintmartin) [inline REVIEW_GUIDELINES.md]: Is it true even for `heritage`?

Comment (scottrigby) [reply REVIEW_GUIDELINES.md]: After further thought, perhaps this paragraph is unnecessary? Now that this PR is separating concerns between `Metadata` (above) and `Deployments and StatefulSets selectors` (below), I'm not sure the goal of this paragraph 🤔 

If the goal is to clarify that by metadata labels we mean `spec.template.metadata.labels`, then perhaps add that to the line above? Something like this:
```diff
-The standard resource metadata should be this:
+The standard resource metadata (`spec.template.metadata.labels`) should be this:
```
Then remove this new paragraph altogether.

Comment (desaintmartin) [reply REVIEW_GUIDELINES.md]: All right! let me propose then:

```diff
-The standard resource metadata should be this:
+The standard resource metadata (`metadata.labels` and `spec.template.metadata.labels`) should be this:
```

Comment (desaintmartin) [reply REVIEW_GUIDELINES.md]: Done.

File: REVIEW_GUIDELINES.md [hunk 3]
```diff
@@ -44,8 +45,27 @@ labels:
   heritage: {{ .Release.Service }}
 ```
 
+If a chart has multiple components, a `component` label should be added (e. g. `component: server`). The resource name should get the component as suffix (e. g. `name: {{ template "myapp.fullname" . }}-server`).
+
 Note that templates have to be namespaced. With Helm 2.7+, `helm create` does this out-of-the-box. The `app` label should use the `name` template, not `fullname` as is still the case with older charts.
 
+### Deployments and StatefulSets selectors
```
Comment (unguiculus) [inline REVIEW_GUIDELINES.md]: We should probably add `DaemonSet` as well.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: REVIEW_GUIDELINES.md (hunk 3 above)
Comment (desaintmartin) [reply REVIEW_GUIDELINES.md]: Thanks, I realized after writing this that DS were affected as well.

**Comment by desaintmartin** [comment]
I've pushed one more commit introducing an upgrade path for selector. I am not sure it is the right place to write it, so feel free to discuss.

File: REVIEW_GUIDELINES.md [hunk 4]
```diff
@@ -44,8 +45,41 @@ labels:
...
+```
+
+If a chart has multiple components, a `component` label should be added (e. g. `component: server`). The resource name should get the component as suffix (e. g. `name: {{ template "myapp.fullname" . }}-server`).
+
+`spec.selector.matchLabels` defined in `Deployments`/`StatefulSets`/`DaemonSets` `>=v1/beta2` **must not** contain `chart` label or any label containing a version of the chart, because the selector is immutable.
+The chart label string contains the version, so if is is specified, whenever the the Chart.yaml version changes, Helm's attempt to change this immutable field would cause the upgrade to fail.
+
+#### Fixing selectors
+
+##### For Deployments, StatefulSets and DaemonSets apps/v1beta1 or extensions/v1beta1
```
Comment (unguiculus) [inline REVIEW_GUIDELINES.md]: Add comma after StatefulSets

File: REVIEW_GUIDELINES.md [hunk 5]
```diff
@@ -44,8 +45,41 @@ labels:
...
+
+- If it does not specify `spec.selector.matchLabel`, set it
+- Remove `chart` label in `spec.selector.matchLabel` if it exists
+- Bump patch version of the Chart
+
+##### For Deployments, StatefulSets and DaemonSets >=apps/v1beta2
```
Comment (unguiculus) [inline REVIEW_GUIDELINES.md]: Add comma after StatefulSets

File: REVIEW_GUIDELINES.md [hunk 6]
```diff
@@ -44,8 +45,41 @@ labels:
...
+
+`spec.selector.matchLabels` defined in `Deployments`/`StatefulSets`/`DaemonSets` `>=v1/beta2` **must not** contain `chart` label or any label containing a version of the chart, because the selector is immutable.
+The chart label string contains the version, so if is is specified, whenever the the Chart.yaml version changes, Helm's attempt to change this immutable field would cause the upgrade to fail.
+
+#### Fixing selectors
+
+##### For Deployments, StatefulSets and DaemonSets apps/v1beta1 or extensions/v1beta1
+
+- If it does not specify `spec.selector.matchLabel`, set it
+- Remove `chart` label in `spec.selector.matchLabel` if it exists
```
Comment (unguiculus) [inline REVIEW_GUIDELINES.md]: matchLabel -> matchLabels

File: REVIEW_GUIDELINES.md [hunk 7]
```diff
@@ -44,8 +45,41 @@ labels:
...
+`spec.selector.matchLabels` must be specified should follow some conventions. The standard selector should be this:
+
+```yaml
+selector:
+  matchLabels:
+    app: {{ template "myapp.name" . }}
+    release: {{ .Release.Name }}
+```
+
+If a chart has multiple components, a `component` label should be added (e. g. `component: server`). The resource name should get the component as suffix (e. g. `name: {{ template "myapp.fullname" . }}-server`).
```
Comment (unguiculus) [inline REVIEW_GUIDELINES.md]: This sentence is copied verbatim from above. It should say that the component label must be added to the selector.

File: REVIEW_GUIDELINES.md [hunk 8]
```diff
@@ -44,8 +45,41 @@ labels:
...
+
+##### For Deployments, StatefulSets and DaemonSets apps/v1beta1 or extensions/v1beta1
+
+- If it does not specify `spec.selector.matchLabel`, set it
+- Remove `chart` label in `spec.selector.matchLabel` if it exists
+- Bump patch version of the Chart
+
+##### For Deployments, StatefulSets and DaemonSets >=apps/v1beta2
+
+- Remove `chart` label in `spec.selector.matchLabel` if it exists
```
Comment (unguiculus) [inline REVIEW_GUIDELINES.md]: matchLabel -> matchLabels

File: REVIEW_GUIDELINES.md [hunk 9]
```diff
@@ -54,7 +88,7 @@ selector:
   release: {{ .Release.Name }}
 ```
 
-If a chart has multiple components, a `component` label should be added (e. g. `component: server`). The resource name should get the component as suffix (e. g. `name: {{ template "myapp.fullname" . }}-server`). The `component` label must be added to label selectors as well.
+If a chart has multiple components, a `component` label should be added (e. g. `component: server`). The resource name should get the component as suffix (e. g. `name: {{ template "myapp.fullname" . }}-server`).
```
Comment (unguiculus) [inline REVIEW_GUIDELINES.md]: Why did you delete this? The component label must be part of the selector.

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by unguiculus** [comment]
Could you also update the persistence section in the guidelines to include `matchLabels`?

**Comment by unguiculus** [comment]
Not sure about PVC. Looks like labels work differently there. I suggested to add `matchLabels` for the PVC in #5987 which causes the PVC to stay pending forever.

File: REVIEW_GUIDELINES.md (hunk 9 above)
Comment (desaintmartin) [reply REVIEW_GUIDELINES.md]: I have separated concerns, so this paragraph does no longer read about selector as there is a dedicated paragraph to it.

**Comment by desaintmartin** [comment]
I think matchLabels for PVC are related to their PV counterpart, which (hopefully) is out of scope. I'd be glad to see confirmation.

I've followed your change requests.

File: REVIEW_GUIDELINES.md [hunk 10]
```diff
@@ -44,8 +45,41 @@ labels:
...
+    app: {{ template "myapp.name" . }}
+    release: {{ .Release.Name }}
+```
+
+If a chart has multiple components, a `component` label should be added to the selector (see above).
+
+`spec.selector.matchLabels` defined in `Deployments`/`StatefulSets`/`DaemonSets` `>=v1/beta2` **must not** contain `chart` label or any label containing a version of the chart, because the selector is immutable.
+The chart label string contains the version, so if is is specified, whenever the the Chart.yaml version changes, Helm's attempt to change this immutable field would cause the upgrade to fail.
+
+#### Fixing selectors
```
Comment (desaintmartin) [inline REVIEW_GUIDELINES.md]: Should this paragraph be here? I'm afraid a dedicated "Breaking change" paragraph might be missed by readers.

File: REVIEW_GUIDELINES.md [hunk 11]
```diff
@@ -44,8 +45,41 @@ labels:
...
+`spec.selector.matchLabels` must be specified should follow some conventions. The standard selector should be this:
+
+```yaml
+selector:
+  matchLabels:
+    app: {{ template "myapp.name" . }}
+    release: {{ .Release.Name }}
+```
+
+If a chart has multiple components, a `component` label should be added to the selector (see above).
```
Comment (desaintmartin) [inline REVIEW_GUIDELINES.md:63]: Is "see above" readable enough?

File: REVIEW_GUIDELINES.md [hunk 12]
```diff
@@ -54,7 +88,19 @@ selector:
   release: {{ .Release.Name }}
 ```
 
-If a chart has multiple components, a `component` label should be added (e. g. `component: server`). The resource name should get the component as suffix (e. g. `name: {{ template "myapp.fullname" . }}-server`). The `component` label must be added to label selectors as well.
+If a chart has multiple components, a `component` label should be added to the selector (see above).
+
+### Persistence labels
```
Comment (desaintmartin) [inline REVIEW_GUIDELINES.md]: Should we add PVC? If so, does anybody have suggestion?

Comment (cpanato) [reply REVIEW_GUIDELINES.md]: if add the matchlabels to the PVC will break and the PVC will never get the PV. only if you set the pv with the lables as well. but I guess we are not generating the PV manifest.

**Comment by cpanato** [comment]
@desaintmartin the command is `/hold` to add the do not merge label

File: REVIEW_GUIDELINES.md (hunk 12 above)
Comment (desaintmartin) [reply REVIEW_GUIDELINES.md]: So should I simply state something like "Unless special needs, you should not set PVC matchLabels because it would prevent automatic PV provisionning"?

Comment (cpanato) [reply REVIEW_GUIDELINES.md]: yes i would say

Comment (desaintmartin) [reply REVIEW_GUIDELINES.md]: Done. Is that okay for you?

Comment (cpanato) [reply REVIEW_GUIDELINES.md]: looks great to me :) thanks!!

File: REVIEW_GUIDELINES.md [hunk 13]
```diff
@@ -54,7 +88,26 @@ selector:
...
+
+### StatefulSet
+
+In case of a `Statefulset`, `spec.volumeClaimTemplates.metadata.labels` must have both `app` and `release` labels, and **must not** contain `chart` label or any label containing a version of the chart, because `spec.volumeClaimTemplates` is immutable.
```
Comment (unguiculus) [inline REVIEW_GUIDELINES.md:97]: I found out that you cannot set `spec.volumeClaimTemplates.metadata.labels`. These labels are automatically set from the StatefulSet's `spec.selector.matchLabels`.

https://github.com/kubernetes/kubernetes/pull/60693
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_utils.go#L152

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: REVIEW_GUIDELINES.md (hunk 13 above)
Comment (desaintmartin) [reply REVIEW_GUIDELINES.md:97]: Oh, right, so we should not put any labels on volumeClaimTemplate?

Comment (desaintmartin) [reply REVIEW_GUIDELINES.md:97]: I'm puzzled here. This is what I do in redis: https://github.com/helm/charts/pull/7686 and it seemed to have an impact (see https://github.com/helm/charts/issues/7803).

**Comment by desaintmartin** [comment]
Should we finish this?

**Comment by mattfarina** [comment]
@desaintmartin @unguiculus @cpanato Can we remove the hold. This LGTM.

**Review by unguiculus — APPROVED** [review]

**Comment by unguiculus** [comment]
I think we can merge this.
